[with Service = tracing::tracing]: Assertion `local_is_initialized()' failed after Tracing is down

Scylla version edc175029712d43aba2ef5845e9ade8375680278
Hit with cql_tracing_test:TestCqlTracing.tracing_shutdown_test in debug mode with https://github.com/bhalevy/scylla-dtest/commit/fbdcc2b74a76d8be0922939926ddb2953062e379

test log:
```
$ CASSANDRA_DIR=$HOME/dev/scylla/build/debug NOSE_PROCESSES=1 nosetests -s -v --process-timeout=1800 cql_tracing_test:TestCqlTracing.tracing_shutdown_test
2019-10-30 09:40:10.811346 going to run tests in parallel
2019-10-30 09:40:10.811428 using the RandomClusterIdAllocator
2019-10-30 09:40:10.845962 cluster ccm directory: /local/home/bhalevy/.dtest/dtest-bwQul7
2019-10-30 09:40:10.846087 Starting Scylla cluster from directory /local/home/bhalevy/dev/scylla/build/debug
2019-10-30 09:40:10.849242 Allocated cluster ID 28: /local/home/bhalevy/.dtest/dtest-bwQul7
2019-10-30 09:40:53.812365 Enable tracing for all CQL requests on node1 and node2...
2019-10-30 09:40:56.673815 CREATE COLUMNFAMILY cf (key varchar PRIMARY KEY, c2 text, c1 text) WITH comment='test cf' AND compression = {} AND read_repair_chance=0.000000
2019-10-30 09:40:58.202978 Populating a table with 500 keys...
2019-10-30 09:41:32.016751 Stopping node1...
2019-10-30 09:42:03.052055 Checking log of node1 for assertions...
2019-10-30 09:42:03.052930 Check that all tracing session have been flushed...
2019-10-30 09:42:04.617291 Start node1...
2019-10-30 09:42:38.232426 Enable tracing for all CQL requests on node1...
2019-10-30 09:42:39.929384 Populating a table with 1000 more keys...
2019-10-30 09:42:39.934211 Wait for 0.563117674109 seconds
2019-10-30 09:42:40.498265 Stopping node2...
2019-10-30 09:42:55.514646 Waiting for insert-thread to complete...
2019-10-30 09:43:21.565713 insertion of 1000 keys is done
2019-10-30 09:43:21.565881 Checking log of node2 for assertions...
2019-10-30 09:43:22.097097 Test failed
2019-10-30 09:43:22.098741 stopping ccm cluster test at: /local/home/bhalevy/.dtest/dtest-bwQul7
2019-10-30 09:43:23.189396 Freeing cluster ID 28: link /local/home/bhalevy/.dtest/28
tracing_shutdown_test (cql_tracing_test.TestCqlTracing) ... FAIL

======================================================================
FAIL: tracing_shutdown_test (cql_tracing_test.TestCqlTracing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/local/home/bhalevy/dev/scylla-dtest/cql_tracing_test.py", line 186, in tracing_shutdown_test
    self.assertEqual(len(match), 0)
AssertionError: 1 != 0
```

[node2.log](https://github.com/scylladb/scylla/files/3787494/node2.log):
```
INFO  2019-10-30 09:42:45,608 [shard 0] tracing - Asked to shut down
INFO  2019-10-30 09:42:45,856 [shard 0] tracing - Tracing is down
scylla: /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/sharded.hh:712: Service& seastar::sharded<T>::local() [with Service = tracing::tracing]: Assertion `local_is_initialized()' failed.
Aborting on shard 0.
Backtrace:
  /lib64/libasan.so.5+0x00000000000b159b
  0x000000000a7a90c4
  0x000000000a7846d2
  0x000000000a611f61
  0x000000000a612059
  0x000000000a6698d7
  0x000000000a6a8422
  0x000000000a6a84c1
  0x00007fb9b08f7e6f
  /lib64/libc.so.6+0x000000000003857e
  /lib64/libc.so.6+0x0000000000022894
  /lib64/libc.so.6+0x0000000000022768
  /lib64/libc.so.6+0x0000000000030a25
  0x0000000000d4a103
  0x0000000000d15624
  0x0000000000d16208
  0x0000000000d15fa5
  0x0000000000d1a4d4
  0x000000000457f80e
  0x00000000057ce31e
  0x0000000005704b13
  0x000000000576cd5c
  0x000000000576ce7e
  0x000000000576cfbb
  0x000000000574e5f1
  0x0000000005735e1a
  0x000000000576d377
  0x0000000000ed86a6
  0x0000000000e72350
  0x0000000000e7244d
  0x0000000000e725ef
  0x0000000000df683d
  0x000000000100b190
  0x000000000a64e68d
  0x000000000a652be5
  0x000000000a6582d6
  0x000000000a484355
  0x000000000a481deb
  0x000000000047fa36
  /lib64/libc.so.6+0x0000000000024412
  0x000000000042a12d

void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/util/backtrace.hh:55
seastar::backtrace_buffer::append_backtrace() at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:680
seastar::print_with_backtrace(seastar::backtrace_buffer&) at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:701
seastar::print_with_backtrace(char const*) at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:708
seastar::sigabrt_action() at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:3378
seastar::install_oneshot_signal_handler<6, &seastar::sigabrt_action>()::{lambda(int, siginfo_t*, void*)#1}::operator()(int, siginfo_t*, void*) const at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:3360
seastar::install_oneshot_signal_handler<6, &seastar::sigabrt_action>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) at /local/home/bhalevy/dev/scylla/seastar/src/core/reactor.cc:3356
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
seastar::sharded<tracing::tracing>::local() at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/sharded.hh:712 (discriminator 1)
tracing::tracing::get_local_tracing_instance() at /local/home/bhalevy/dev/scylla/./tracing/tracing.hh:424
tracing::one_session_records::set_pending_for_write() at /local/home/bhalevy/dev/scylla/./tracing/tracing.hh:668
tracing::tracing::schedule_for_write(seastar::lw_shared_ptr<tracing::one_session_records>) at /local/home/bhalevy/dev/scylla/./tracing/tracing.hh:573
tracing::trace_state::trace_internal(seastar::basic_sstring<char, unsigned int, 15u, true>) at /local/home/bhalevy/dev/scylla/tracing/trace_state.hh:557
void tracing::trace_state::trace<gms::inet_address const&>(char const*, gms::inet_address const&) at /local/home/bhalevy/dev/scylla/./tracing/trace_state.hh:569 (discriminator 1)
void tracing::trace<char const (&) [22], gms::inet_address const&>(tracing::trace_state_ptr const&, char const (&) [22], gms::inet_address const&) at /local/home/bhalevy/dev/scylla/./tracing/trace_state.hh:696
db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}::operator()({lambda()#1}) const at /local/home/bhalevy/dev/scylla/db/hints/manager.cc:176
seastar::apply_helper<db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, std::tuple<{lambda()#1}&&>&&, std::integer_sequence<unsigned long, 0ul> >::apply(db::rp_handle, std::tuple) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/apply.hh:36
auto seastar::apply<db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, {lambda()#1}&&>(db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, std::tuple<{lambda()#1}&&>&&) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/apply.hh:44
seastar::future<> seastar::do_void_futurize_helper<void>::apply_tuple<db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, {lambda()#1}&&>(db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, std::tuple<{lambda()#1}&&>&&) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/future.hh:1489
seastar::future<> seastar::futurize<void>::apply<db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, {lambda()#1}&&>(db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&, std::tuple<{lambda()#1}&&>&&) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/future.hh:1523
_ZZN7seastar6futureIJN2db9rp_handleEEE4thenIZZZNS1_5hints7manager23end_point_hints_manager10store_hintENS_13lw_shared_ptrIK6schemaEENS8_IK15frozen_mutationEEN7tracing15trace_state_ptrEENUlvE_clEvENUlvE_clEvEUlS2_E0_NS0_IJEEEEET0_OT_ENUlOS2_E_clESO_ at /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/future.hh:1074
seastar::noncopyable_function<seastar::future<> (db::rp_handle&&)>::direct_vtable_for<seastar::future<db::rp_handle>::then<db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}, seastar::future<> >(db::hints::manager::end_point_hints_manager::store_hint(seastar::lw_shared_ptr<schema const>, seastar::lw_shared_ptr<frozen_mutation const>, tracing::trace_state_ptr)::{lambda()#1}::operator()()::{lambda()#1}::operator()()::{lambda(db::rp_handle)#2}&&)::{lambda(db::rp_handle&&)#1}>::call(seastar::noncopyable_function<seastar::future<> (db::rp_handle&&)> const*, db::rp_handle&&) at /local/home/bhalevy/dev/scylla/seastar/include/seastar/util/noncopyable_function.hh:99
seastar::noncopyable_function<seastar::future<> (db::rp_handle&&)>::operator()(db::rp_handle&&) const at /local/home/bhalevy/dev/scylla/seastar/include/seastar/util/noncopyable_function.hh:182
```

This is reproducible most of the times on my machine.

--- a/tracing/tracing.cc
+++ b/tracing/tracing.cc
@@ -206,8 +206,9 @@ void tracing::set_trace_probability(double p) {
 }
 
 one_session_records::one_session_records()
-    : backend_state_ptr(tracing::get_local_tracing_instance().allocate_backend_session_state())
-    , budget_ptr(tracing::get_local_tracing_instance().get_cached_records_ptr()) {}
+    : _local_tracing_ptr(tracing::get_local_tracing_instance().shared_from_this())
+    , backend_state_ptr(_local_tracing_ptr->allocate_backend_session_state())
+    , budget_ptr(_local_tracing_ptr->get_cached_records_ptr()) {}
 
 std::ostream& operator<<(std::ostream& os, const span_id& id) {
     return os << id.get_id();

--- a/tracing/tracing.hh
+++ b/tracing/tracing.hh
@@ -240,6 +240,8 @@ public:
 };
 
 class one_session_records {
+private:
+    shared_ptr<tracing> _local_tracing_ptr;
 public:
     utils::UUID session_id;
     session_record session_rec;
@@ -665,7 +667,7 @@ private:
 
 void one_session_records::set_pending_for_write() {
     _is_pending_for_write = true;
-    budget_ptr = tracing::get_local_tracing_instance().get_pending_records_ptr();
+    budget_ptr = _local_tracing_ptr->get_pending_records_ptr();
 }
 
 void one_session_records::data_consumed() {
@@ -674,7 +676,7 @@ void one_session_records::data_consumed() {
     }
 
     _is_pending_for_write = false;
-    budget_ptr = tracing::get_local_tracing_instance().get_cached_records_ptr();
+    budget_ptr = _local_tracing_ptr->get_cached_records_ptr();
 }
 
 inline span_id span_id::make_span_id() {

